LocusWalkerSpark drops intervals with no reads

tomwhite opened the pull request:
This PR fixes #3823 opened by @LeeTL1220 

There are several series of commits applied here:

1. The first set of commits rebase `ll_CollectAllelicCountsSpark` on `master`.
2. Then there's a commit (https://github.com/broadinstitute/gatk/commit/21e1dcfb88fc6543f6ba3e6095eba512a33f9f8d) to fix up some changes to make `CollectAllelicCountsSpark` work on `master`.
3. The next set of commits applies the changes from  #5127 and #5221 (since they have not yet been merged) to make passing the reference in Spark more efficient. These changes are needed for the fix below to work.
4. The actual fix is in https://github.com/broadinstitute/gatk/commit/3326b9093246ff6fb51ad5537951b4646411d80f.
5. There's also a new test for `ExampleLocusWalkerSpark` in addition to the one for `CollectAllelicCountsSpark`.

The problem was that `emitEmptyLoci()` in `LocusWalkerSpark` was not working properly. Any intervals that didn't overlap with reads were being dropped, which meant that those loci were not being passed to the `LocusWalkerSpark` subclass.

LeeTL1220 commented:
@tomwhite The other changes should probably be reviewed by someone else on the engine team.

droazen commented:
This should not be merged until #5127 and #5221 have gone in first -- treating as blocked until then

tomwhite commented:
@LeeTL1220 thanks for the review! I'll work with the engine team to get the other parts reviewed.

jamesemery commented:
@tomwhite Could this branch be disentangled from `ll_CollectAllelicCountsSpark` since the actual fix is in 3326b90 we should probably just pull that out into its own branch and push that through with `ll_CollectAllelicCountsSpark` which doesn't seem directly related.

tomwhite commented:
Now that  #5127 and #5221 have gone in I've rebased this. Note that it will fail until the fix (which I moved to #5248 on @jamesemery's suggestion) has gone in.
